### PR TITLE
SNOW-958008: fix local test division datatypes

### PR DESCRIPTION
**Pull request by sfc-gh-yixie**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/types.py [hunk 1]
```diff
@@ -80,7 +80,7 @@ def __init__(self, length: Optional[int] = None) -> None:
         self.length = length
 
     def __repr__(self) -> str:
-        if self.length:
+        if self.length and self.length != self._MAX_LENGTH:
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/types.py:83]: @sfc-gh-aalam Are you good with this change?

File: src/snowflake/snowpark/_internal/analyzer/expression.py [hunk 2]
```diff
@@ -214,6 +214,8 @@ def __init__(self, value: Any, datatype: Optional[DataType] = None) -> None:
             self.datatype = datatype
         else:
             self.datatype = infer_type(value)
+            if isinstance(self.datatype, StringType):
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/_internal/analyzer/expression.py:217]: @sfc-gh-aalam Are you good with this change?

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/expression.py:217]: yes. approved the other PR which adds this change

File: src/snowflake/snowpark/mock/plan.py [hunk 3]
```diff
@@ -946,15 +946,15 @@ def calculate_expression(
         )
         return ColumnEmulator(
             data=[bool(data is None) for data in child_column],
-            sf_type=ColumnType(BooleanType(), False),
+            sf_type=ColumnType(BooleanType(), True),
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/plan.py:949]: We observe that boolean type columns are always nullable from live connection.

File: src/snowflake/snowpark/mock/plan.py [hunk 4]
```diff
@@ -989,6 +989,8 @@ def calculate_expression(
     if isinstance(exp, BinaryExpression):
         left = calculate_expression(exp.left, input_data, analyzer, expr_to_alias)
         right = calculate_expression(exp.right, input_data, analyzer, expr_to_alias)
+        # TODO: Address mixed type calculation here. For instance Snowflake allows to add a date to a number, but
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/plan.py:992]: This should be enabled by type-coercion.

File: src/snowflake/snowpark/mock/snowflake_data_type.py [hunk 5]
```diff
@@ -182,7 +191,7 @@ def calculate_type(c1: ColumnType, c2: Optional[ColumnType], op: Union[str]):
     elif isinstance(t1, (FloatType, DoubleType)) or isinstance(
         t2, (FloatType, DoubleType)
     ):
-        return ColumnType(DoubleType(), nullable)
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/snowflake_data_type.py:185]: We could switch to using `normalize_output_sf_type`

File: src/snowflake/snowpark/mock/snowflake_data_type.py [hunk 6]
```diff
@@ -435,25 +446,28 @@ def __round__(self, n=None):
...
+        return other.__truediv__(self)
 
     def __truediv__(self, other):
         result = super().__truediv__(other)
-        result.sf_type = calculate_type(self.sf_type, other.sf_type, op="/")
+        sf_type = calculate_type(self.sf_type, other.sf_type, op="/")
+        if isinstance(sf_type.datatype, DecimalType):
+            result = result.astype("double").round(sf_type.datatype.scale)
+        elif isinstance(sf_type.datatype, (FloatType, DoubleType)):
+            result = result.astype("double").round(16)
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/snowflake_data_type.py:461]: We observe that Snowflake returns 16 digits after the floating point. cc: @sfc-gh-yixie

File: tests/integ/test_datatypes.py [hunk 7]
```diff
@@ -36,13 +35,14 @@ def test_basic_filter():
             [
                 StructField("A", LongType(), nullable=False),
                 StructField("B", LongType(), nullable=False),
-                StructField("C", StringType(), nullable=False),
+                StructField("C", StringType(16777216), nullable=False),
```
Comment (sfc-gh-stan) [inline tests/integ/test_datatypes.py:38]: Do we need to specify this size here?

File: tests/integ/test_datatypes.py [hunk 8]
```diff
@@ -156,9 +159,27 @@ def test_divide_basic():
             ]
         )
     )
+    Utils.check_answer(df, [Row(1, 0.3333333, 0.7333333333333334)])
```
Comment (sfc-gh-stan) [inline tests/integ/test_datatypes.py:162]: Should we use `Decimal("0.3333333")` instead?

File: tests/integ/test_datatypes.py [hunk 9]
```diff
@@ -156,9 +159,27 @@ def test_divide_basic():
...
+        schema=StructType(
+            [StructField("a", DoubleType()), StructField("b", DoubleType())]
+        ),
+    )
+    df = df.select([df["a"] / df["b"]])
+    Utils.check_answer(df, [Row(0.8461538461538461)])
 
+    df2 = session.create_dataframe([[11, 13]], schema=["a", "b"])
+    df2 = df2.select([df2["a"] / df2["b"]])
+    Utils.check_answer(df2, [Row(0.846154)])
```
Comment (sfc-gh-stan) [inline tests/integ/test_datatypes.py:178]: Decimal("0.846154")

File: tests/integ/test_datatypes.py [hunk 10]
```diff
@@ -262,39 +283,54 @@ def test_unary_ops_bool():
...
-                StructField('REGEXP("A")', BooleanType(), nullable=False),
+                StructField('"""A"" LIKE \'V%\'"', BooleanType(), nullable=True),
+                StructField('"""A"" REGEXP \'V\'"', BooleanType(), nullable=True),
             ]
         )
     )
 
 
-def test_filter():
+@pytest.mark.skip("Cast is not implemented yet.")
```
Comment (sfc-gh-stan) [inline tests/integ/test_datatypes.py:320]: Let's remove these.

File: src/snowflake/snowpark/mock/snowflake_data_type.py [hunk 11]
```diff
@@ -435,25 +446,28 @@ def __round__(self, n=None):
 
     def __rpow__(self, other):
         result = super().__rpow__(other)
-        result.sf_type = ColumnType(DoubleType(), self.sf_type.nullable)
+        result.sf_type = ColumnType(FloatType(), True)
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/snowflake_data_type.py:449]: same comment as Sophie's to use `normalize_output_sf_type`

File: src/snowflake/snowpark/mock/snowflake_data_type.py [hunk 12]
```diff
@@ -339,57 +348,59 @@ def __bool__(self):
...
         result = super().__xor__(other)
-        result.sf_type = ColumnType(BooleanType(), self.sf_type.nullable)
+        result.sf_type = ColumnType(BooleanType(), True)
         return result
 
     def __pow__(self, power):
         result = super().__pow__(power)
-        result.sf_type = ColumnType(DoubleType(), self.sf_type.nullable)
+        result.sf_type = ColumnType(
+            FloatType(), self.sf_type.nullable or power.sf_type.nullable
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/snowflake_data_type.py:372]: are all the `FloatType` in the code usage ultimately be converted to `DoubleType` via `normalize_output_sf_type` at some point?
and from user perspective will they get a chance to see the `FloatType` at all?

it seems the two types are interchangeable and we want `DoubleType` to be in the final output. can we just use DoubleType everywhere?

**Comment by sfc-gh-stan** [comment]
The code changes LGTM, let's address the test failures in Test Local Testing Module macos merge gates.

File: src/snowflake/snowpark/mock/plan.py [hunk 13]
```diff
@@ -1661,7 +1670,9 @@ def calculate_expression(
                     ).sf_type
                     if not calculated_sf_type:
                         calculated_sf_type = cur_windows_sf_type
-                    elif calculated_sf_type.datatype != cur_windows_sf_type.datatype:
+                    elif not is_datatype_compatible(
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/plan.py:1673]: This basically allows coercion to happen but doesn't set the result type properly?

Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/mock/plan.py]: i removed it

File: src/snowflake/snowpark/mock/functions.py [hunk 14]
```diff
@@ -745,16 +745,27 @@ def mock_iff(condition: ColumnEmulator, expr1: ColumnEmulator, expr2: ColumnEmul
...
-            else expr2.sf_type.datatype
-        )
+        if isinstance(expr1.sf_type.datatype, StringType) and isinstance(
+            expr2.sf_type.datatype, StringType
+        ):
+            l1 = expr1.sf_type.datatype.length or StringType._MAX_LENGTH
+            l2 = expr2.sf_type.datatype.length or StringType._MAX_LENGTH
+            sf_data_type = StringType(max(l1, l2))
+        else:
+            sf_data_type = (
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/functions.py:764]: Are we not using `calculate_datatype` to compute the return type?

**Review by sfc-gh-aling — APPROVED** [review]

**Pull request merged**